### ISSUE TITLE
UKI + systemctl kexec

### ISSUE THREAD

**Issue by Cydox** [systemd/systemd#28538]
### Describe the solution you'd like

Currently `systemctl kexec` does not work when using UKIs. It should be possible to extract the components of the UKI and hand that over to `kexec` in a way that it can understand.

I am willing to work on this.

### Describe alternatives you've considered

Alternatively `kexec` can be taught how to deal with UKIs itself. There seems to have been some discussion of this [here](https://www.spinics.net/lists/kexec/msg30939.html), but it looks like it hasn't gone anywhere.

### The systemd version you checked that didn't have the feature you are asking for

252

**Comment by bluca** [comment]
The problem with extracting components is signature checks - IIRC when lockdown is enabled, you can only kexec load a signed image. But we don't sign individual components intentionally in a UKI, we sign the envelope. So I think there might not really be an alternative to teaching kexec to load and verify a UKI?

**Comment by Cydox** [comment]
@bluca You're right, this approach would only work without lockdown mode / secure boot disabled which kinda defeats the purpose of using UKIs.

I assume the signature is checked by the kernel and not the userspace kexec utility, right?

**Comment by bluca** [comment]
Yes, there is a specific flag you pass to kexec load, and it goes through a different path that makes the kernel check the signature - I do not remember exactly what they are called, just going from memory here

**Comment by Cydox** [comment]
Alright, it's the version of the syscall where a file descriptor is passed in. Kernel loads that file and checks the signature, and then sets up the purgatory to boot the new kernel. As far as I can tell this is only implemented for bzImages (`arch/x86/kernel/- kexec-bzimage64.c`).

So I think this feature will require support from the kernel. I wonder if just loading the relevant sections from the UKI in `kexec` is generic enough. Maybe it's possible to implement kexec for loading generic efi applications that are signed. Don't know if we can launch another efi application after boot services are stopped (will look through the UEFI spec later). Any other input on this?

**Comment by poettering** [comment]
In the long run kexec would have to replicate what uefi/sd-stub currently do at boot:

1. validate the whole UKI PE signature
2. take apart the PE sections, as sd-stub does
3. look for sysext + creds + addons at some place like sd-stub does
4. measure the parts as sd-stub does to TPM2
5. write a log about what it just measured, and somehow pass it over to the new kernel
6. then invoke the kernel with the various components

This of course is quite involved. And probably will grow eventually. For example, once "systemd-pcrlock" lands we should probably also generate a new PCR signature for the expected PCR state after the kexec, so that access to TPM objects hat are protected by signed PCR policies can be retained by the new kernels.

To me, I kinda sense that we should probably provide a wrapper `systemd-kexec` or so sooner or later, that does all these steps. And that is developed to closely follow what sd-stub does.

**Comment by bluca** [comment]
The problem is that it necessarily needs to happen in the kernel, or else you'd be allowing to load an unsigned payload, which can't even work when secure boot is enabled due to lockdown (for very good reasons)

**Comment by poettering** [comment]
i don't think it necessarily has to. if you buy into the lockdown thing, then you only care about the kernel's own integrity not userspace for that, hence it's enough if the the kernel checks some sig on the inner kernel, and leaves the additional checking of the other stuff to userspace. thus you'd sign + validate the kernel twice: the "inner" kernel is validated in kernel mode, and the "outer" envelope" is validated in userspace.

**Comment by bluca** [comment]
Having to sign both inner and outer is a bit of an annoyance, although doable. But losing the property that we load trusted initrds is a big downside. If our userspace tool is allowed to kexec an unsigned initrd, then anything is allowed to kexec an unsigned initrd, which is a bummer.

**Comment by poettering** [comment]
signed initrds contradict with sysext, credentials, which are placed in on-the-fly initrds after all...

tbh I am not too concerned by this. as long as the kernel measures the whole initrd it's ok if it comes with a weaker validation, because you then can lock secrets to the expected initrd, and do attestation. 

i mean, this is my general thinking: measured boot and locking keys to it is a million times better than trying to restrict what is booted. i.e. I am nowadays firmly in the camp: fuck SecureBoot, yay MeasuredBoot. If you have the latter then you don't really need the former so much.

current kernels measure the full initrd they got when booted via uefi mode. What's missing is that they do the same when booted via kexec, without a chance for userspace to interfere. (or does that already exist?) if so, we should have enough.

**Comment by Cydox** [comment]
> Having to sign both inner and outer is a bit of an annoyance, although doable.

This relates to the email I sent. Can I just explain it here? At least as it relates to this issue.

**Comment by Cydox** [comment]
> tbh I am not too concerned by this. as long as the kernel measures the whole initrd it's ok if it comes with a weaker validation, because you then can lock secrets to the expected initrd, and do attestation.
> 
> i mean, this is my general thinking: measured boot and locking keys to it is a million times better than trying to restrict what is booted. i.e. I am nowadays firmly in the camp: fuck SecureBoot, yay MeasuredBoot. If you have the latter then you don't really need the former so much.

In this case UKIs are not needed at all. With a normal kernel image + initrd + cmdline from sd-boot we get all the measurements we need.


> current kernels measure the full initrd they got when booted via uefi mode. What's missing is that they do the same when booted via kexec, without a chance for userspace to interfere. (or does that already exist?) if so, we should have enough.

Tested this on an Arch Linux VM right now. Kexecing into an updated kernel does not change any PCR values. So at least with the kernel config of Arch nothing gets measured on kexec.

But how would this work exactly. We can't reset the PCRs (right?). So the PCR values would depend on the whole chain of previously booted kernels / initrds / cmdlines. Do you want the PCR signatures to be generated on the fly (meaning the key for signing has to live somewhere on the device)?

**Comment by poettering** [comment]
> In this case UKIs are not needed at all. With a normal kernel image + initrd + cmdline from sd-boot we get all the measurements we need.

well, i think there's value in restricting what a system can boot into, but that's just so that noone can use your device outside of the intended usecase. Which is quite different from goal of protecting secrets. But yeah, in some scenarios it migh tbe worth just ditching SB entirely, and just caring for MB.



> But how would this work exactly. We can't reset the PCRs (right?). 

Correct. (There's one debug PCR which you can reset, but generally no, you don#t get to reset anything without rebooting).

> So the PCR values would depend on the whole chain of previously booted kernels / initrds / cmdlines. Do you want the PCR signatures to be generated on the fly (meaning the key for signing has to live somewhere on the device)?

So the model I am going for with the new "pcrlock" stuff I am working on is to maintain a signing key pair in the TPM itself that can be used to sign new access policies to itself if a previous signed access policy can be provided. The idea would then be that we lock down access to this key with all PCRs we possibly can. Whenever we expect changes to it we'll sign a new policy matching the expectations, and invalidate the old policies (via a tpm2 counter). If we expect a change we cannot predict (i.e. firmware upgrade we don't control), then we can generate a "lose" policy that drops a specific PCR from the policy for one boot and restricts it back on the next.

This model would then be extensible to the kexec key: when we are about to reboot, we sign a new policy with the key, that makes it accessible on the next boot under the right conditions (i.e. if the PCRs are updated the exact way we expect them to be). We'd then pass this signature to the new boot where it can be used. When that system then does another kexec reboot it goes through the same dance.

So in this model the OS would never get direct access to the signature key (because it is stored in non-extractable way in the TPM), but we'd basically pass around a "token" (specifically: a signed PCR policy) from one boot to the next that grants access in the expected environment and state, but nothing else. For hard reboots we would invalidate the old policies via a tpm2 counter, and for kexec reboots we'd probably invalidate them by measuring a nonce into some PCR. Every boot would gain the ability to generate arbitrary tokens for arbitrary states though. This means if you exploit a system, you could possibly be able to ask for a token with all protections off, and then reboot and use the token to do whatever you want, all from userspace. This would leave only a minimal window of opportunity though as the expectation is that early from the initrd we'd unconditionally lock down the policy again to what the vendor said. So basically, in this model if you exploit userspace you might be able relax policies for a short moment of the subsequent boot, but once regular userspace took over again you lost it all again.

Of course, ideally we'd have an even tighter model, but I think it's pretty good given that PCRs are so hard to predict given firmware upgrades don't tell use PCR changes clearly ahead f time. we have to work with what we get.

**Comment by Cydox** [comment]
I think there is value in supporting a simpler model that does not extend trust to userspace.

I have a proof of concept implementation of in-kernel loading of UKIs for kexec at: [https://github.com/Cydox/linux/commits/kexec-uki](https://github.com/Cydox/linux/commits/kexec-uki)

The kernel just loads the entire UKI, validates it's signature (if in lockdown mode or the kernel is compiled with CONFIG_KEXEC_SIG_FORCE), then it extracts the `.linux`, `.initrd`, and `.cmdline` sections from the UKI calls the normal bzImage kexec loader from there.

This will also require changes to the userspace `kexec` cli, as it currently complains about not understanding the format. So I wrote a very simple userspace tool for testing purposes that directly does the syscall. It is at [https://github.com/Cydox/kexec-test](https://github.com/Cydox/kexec-test)

For anyone interesting in testing you'll need to:
1. Compile your kernel with my changes
2. Build my userspace `kexec-test` tool
3. Load your UKI using `kexec-test <path to your UKI>`
4. Call `systemctl kexec` as normal. It will use the previously loaded UKI

The implementation needs some work still. It needs code for detecting if an image is a valid UKI. The code used for parsing pe files was taken from another place in the kernel and slightly modified. So this should be merged back with the other code to reduce duplication.

**Comment by bluca** [comment]
> Different question: Can the `systemd-stub` run after EFI boot services are terminated?
> 
> Asking because I'm wondering if instead of extracting the kernel, initrd, and cmdline and kexecing them like normal the kernel should kexec into `systemd-stub`.

I don't think so, also I don't think you can jump back into EFI mode?

**Comment by bluca** [comment]
> Getting a little more high level again. I think it's a valid usecase to just wanne use UKI for secure boot signing without doing any measured boot at all. The in-kernel approach is in my opinion the obvious choice for that. Having in-kernel UKI validation and loading does not interfere with systemd's ability to do something else in the measured boot case.

I fully agree, and like your approach. There is huge value in being able to support kexec'ing a UKI and in fact (if I ever switch to EFI at work) I will need exactly that too. How likely would it be to have such changes accepted in the kernel? Bearing in mind that we fully intend for the UKI format to be an open and stable specification, that anybody can support and rely on. We might add things (like addons and so on) but I think the current format is stable.

Regarding addons, once the tools learn to pass a PE to kexec, and the kernel learns how to unpack it into its components, it shouldn't be too much of a stretch to add support for loading multiple PEs for command line extensions, which is the currently supported addon, and initrd in the future? Once validated, multiple initrds can be easily concatenated. So the kexec tool would figure out what to load from where, and the kernel would just unpack and use everything. This should work in principle, right?

**Comment by keycommander** [comment]
Very interesting compilation of work.

**Comment by poettering** [comment]
> The model with on-the-fly generated PCR signatures on the other hand needs to trust userspace. As you say here you want to have the next kernel measure the initrd/cmdline without ability of userspace to interfere. However userspace is generating the expected PCR value that is then signed by the TPM. So why is it important for userspace not to interfere in the measurement if it's generating this value in the first place?

Oh, for my pcrlock scheme the kernel's own initrd measurement doesn't matter. I was more thinking of a remote attestation scheme, independently of that: if some orchestrator wants to prove the node is in order it should be able to rely on measurements the kernel did, not on those userspace did.

In a way measuring the initrd measurements the kernel does in UEFI mode and the initrd measurements sd-stub does (and I think we should do before kexec in userspace) are redundant, since they cover the same bits and bytes. However, in another way they are not: to the kernel "the initrd" is the concatenation of various bits we compile in sd-stub, such as credentials, sysexts, actual initrds and in the future probably more stuff (confexts, …). In sd-stub/userspace otoh these parts are separate concepts and we measure them to different PCRs, since some of it is more "kernel parameters" and others is actually code. And of course, if you take an outside view, who does the measurements matters too: kernelspace or userspace, for the purpose of remote attestation.

Does that make any sense?

**Comment by poettering** [comment]
> Different question: Can the `systemd-stub` run after EFI boot services are terminated?

No. It's strictly code that only works pre-ExitBootServices().

**Comment by Cydox** [comment]
There are (at least) three distinct usecases:

- **(A)**: Secure boot with UKI, no measured boot.
- **(B)**: Secure boot with UKI + measuerd boot for local secrets (e.g. disk encryption keys in TPM)
- **(C)**: Secure boot with UKI + measured boot for remote attestation

## A

This is where the in-kernel approach seems most appropriate.

> How likely would it be to have such changes accepted in the kernel?

Never contributed to the kernel so I don't know. However I hope that it will be considered a stable enough format to be accepted. I can polish up my patch this week and/or next week and put it out on the mailing list.

What is the official place where the format is documented? [UAPI bootloader spec](https://uapi-group.org/specifications/specs/boot_loader_specification/#type-2-efi-unified-kernel-images) refers to the man page for `systemd-stub`.

> Regarding addons, once the tools learn to pass a PE to kexec, and the kernel learns how to unpack it into its components, it shouldn't be too much of a stretch to add support for loading multiple PEs for command line extensions, which is the currently supported addon, and initrd in the future? Once validated, multiple initrds can be easily concatenated. So the kexec tool would figure out what to load from where, and the kernel would just unpack and use everything. This should work in principle, right?

I think in principle this can work, however will require a new syscall (or a new flag for the current syscall), as the current one only takes two file descriptors (one for kernel other for initrd, my patch is using the kernel one for UKIs). And calling it multiple times simply replaces the  previously loaded kernel. I think adding a new flag to the syscall that results in the kernel not replacing the previous UKI but instead loading the new file as an extension is a good approach. This could come as an incremental improvement after getting initial support.

## B

There's more problems here. Since the disk encryption has to be unlocked again, we basically have two options to support this.

1. Don't re-measure and pass on the PCR signatures. Open Question: How do we pass them on. One solution is for the kernel to retrieve them (by what means?) and insert them into the initrd before the kexec (similar to what sd-boot does if I'm not mistaken). This would require more support from the kernel. Getting the PCR signatures to the kernel would probably require a new file somewhere in `sysfs` I assume.
2. Sign the new expected PCR values on the machine and pass those on. This is the model and usecase I was imagining when I was reading @poettering's [comment](https://github.com/systemd/systemd/issues/28538#issuecomment-1669636217) about the new `pcrlock` stuff. I did not think of usecase C at the time.

IMO option 1 is the better one for this usecase, as discussed earlier.

## C

> Oh, for my pcrlock scheme the kernel's own initrd measurement doesn't matter. I was more thinking of a remote attestation scheme, independently of that: if some orchestrator wants to prove the node is in order it should be able to rely on measurements the kernel did, not on those userspace did.

I think this usecase is where the `pcrlock` scheme can really shine. Even if the model is not really tight from the perspective of local secrets as the next PCR values would need to be signed on the fly locally, the remote attestor is in a different position. Getting both the PCR values and the TPM eventlog the remote attestor can verify the entire chain. You can do all sorts of nice stuff with this like only releasing secrets to machines that are running an up-to-date kernel, etc.

I believe there is some work in this area by Google/Matthew Garrett.

**Comment by bluca** [comment]
> What is the official place where the format is documented? [UAPI bootloader spec](https://uapi-group.org/specifications/specs/boot_loader_specification/#type-2-efi-unified-kernel-images) refers to the man page for `systemd-stub`.

https://uapi-group.org/specifications/specs/unified_kernel_image/

**Comment by Cydox** [comment]
Alright so I have some more thoughts and I think I came up with a plan that can cover everything.

Steps for kernel support:
1. Basic support for kexec'ing UKIs (see proof of concept implementation)
2. support for addons via new flag for `kexec_file_load`
3. support for passing PCR signatures to the next kernel via `initrd`. These signatures can be passed to the kernel via `sysfs`. I thought of just passing them using an addon, however in that case they would have to be signed by the secure boot or kernel signing keys.
4. kernel implementation of TPM re-measurement. This will have to be optional (maybe decided via another flag to `kexec_file_load`).

This gives `systemd` all kernel interfaces to cover usecases A, B, and C.

How `systemd` could handle `systemctl kexec` in the different usecases:
- **A**: No change necessary. Just call `kexec` on the UKI and as long as step 1 is implemented everything works
- **B**: Same as A, but pass on the current PCR signatures to the kernel. Requires step 1, and 3. Second option: compute new expected PCR values (and have them signed by keys stored on the TPM), and pass those on to the kernel for `kexec` and tell the kernel to re-measure. Requires step 1, 3, and 4.
- **C**: Basically the implementation described as second option for B.


There's quite a lot of kernel work involved in making everything work, however it is incremental. Just getting step 1 in the kernel will enable usecase A. Does anyone have an idea how likely it is to get this into the kernel? I feel most optimistic about step 1 and least optimistic about step 4 (but I don't have experience with getting stuff in the kernel).

The other option is to go more userspace oriented and get this stuff working just with changes to `systemd`. The problem with that route is that it will require signing the kernel image inside the UKI, which in my opinion defeats the purpose of UKIs as anyone can then pull it apart and boot a different initrd/cmdline. So I think the best solution for this is the in-kernel option.

I'm willing to put in work to get this into the kernel starting with step 1.


@poettering and @bluca please critique.

**Comment by bluca** [comment]
Starting with step 1 sounds like a good plan to me, and then we can see the reaction to that and take it from there

**Comment by Cydox** [comment]
Just mailed the patch to the mailing list.
[https://lore.kernel.org/all/20230909161851.223627-1-kernel@jfarr.cc/T/](https://lore.kernel.org/all/20230909161851.223627-1-kernel@jfarr.cc/T/)

**Comment by Cydox** [comment]
Version 2 of the patch: https://lore.kernel.org/kexec/20230911052535.335770-1-kernel@jfarr.cc/T/

**Comment by Cydox** [comment]
>    **B**: Same as A, but pass on the current PCR signatures to the kernel. Requires step 1, and 3. Second option: compute
> new expected PCR values (and have them signed by keys stored on the TPM), and pass those on to the kernel for `kexec` and > tell the kernel to re-measure. Requires step 1, 3, and 4.


Option 1 here does not work like that, because of systemd-pcrphase.service. Unless you sign a policy for phases beyond leaving the initrd for your rootfs, but obviously this opens up some possible attacks (like tricking the kernel/initramfs to mount a different rootfs maybe, then using root inside that rootfs to mount the real rootfs later).

So instead the key for the rootfs would have to be passed to the new kernel in an on-the-fly initrd maybe. I hope the initrd is not accessibly by userspace later on (after execution is passed to rootfs).

**Comment by ervinpopescu** [comment]
@Cydox I tried compiling the kernel with your changes from [c99fde](https://github.com/Cydox/linux/commit/c99fde8248c8f25cab222297d4847ef402b96c86) onwards, but the `kexec-test` program seems to fail loading the UKI, as the return value is -1 and `systemctl kexec` afterwards still loads the wrong UKI. Please note I cherry picked the commits to the latest kernel commit. I will retry with your branch and let you know my results.

```console
$ uname -r
6.6.0-rc7-kexec-test-00148-g41a115b45c56
```

**Comment by Cydox** [comment]
@ervinpopescu
`dmesg` should have more info on what went wrong. Can you post the end of that log (should be the last 10 or so lines).

**Comment by ervinpopescu** [comment]
The [`dmesg` output](https://termbin.com/cdwk) does not seem to be related/useful... Tried with [`journalctl`](https://termbin.com/5607) too, no dice. Currently on `6.5.0-next-20230907-kexec-test-old-14164-g2fa3ebb77fa8`, `kexec-test` still not loading the UKI.

**Comment by ervinpopescu** [comment]
Finally worked, I made a stoopid rookie mistake...
